### ISSUE TITLE
RayOutOfMemoryError when running stablizing_highway example

### ISSUE THREAD

**Issue by buechelmartin** [flow-project/flow#651]
Although this might be an issue of ray, since triggered from flow code, I decided to open this as a bug report here.
There seem to be a few ongoing issues with ray itself, where the origin of the problem most likely lies.
The purpuse of this bug report should be to make sure that eventually, flow will support future versions of ray, with all the correct configurations necessary to avoid this bug.

Here are related issues I found in rllib/ray:

https://github.com/ray-project/ray/issues/5211

https://github.com/ray-project/ray/issues/4877

https://github.com/ray-project/ray/issues/4498

https://github.com/ray-project/ray/issues/4357

https://github.com/ray-project/ray/pull/3323

Comment: for the moment, I would highly appreciate to get any hints for how to avoid this from happening, since it hinders me from running any experiments over enough iterations.

## Bug Description

Running examples/rllib/stabilizing_highway.py leads to the failure below, causing the experiment to be aborted.

## Bug Reproduce
Unfortunately, I cannot always reproduce this, since many times my experiment will just stop without any error message. I'm trying to provide more insight on this (presumably different) issue in another issue/bug report as soon as I got this. 
Having tried the example:
examples/rllib/stabilizing_highway.py
with several parameter combinations, this time it happened with:

```
# time horizon of a single rollout
HORIZON = 750
# number of rollouts per training iteration
N_ROLLOUTS = 50
# number of parallel workers
N_CPUS = 1
   config["model"].update({"fcnet_hiddens": [100, 50, 25]})
```

somewhere between checkpoint 120 and 140, but I assume this is depending on my memory.

My feeling was that this happens more often when running more parallel workers, thats why I tried out running only one.

(Although, this leads to 2 CPUs being initialized in the example code due to:

`ray.init(num_cpus=N_CPUS + 1, redirect_output=False)`

but I will raise another issue to clarify why this is necessary.

I tried reducing the Ray object store size by setting `object_store_memory` when calling `ray.init`, but this had no effect. Anyway, I agree with one of the comments in a linked issue, that the memory should used should not be growing during the experiments.

## Output Log

```
Loading configuration... done.
Success.
 Starting SUMO on port 49285
 Cfg file: /home/buechel/gitfolders/flowfolder/flow/core/kernel/scenario/debug/cfg/stabilizing_open_network_merges_20190716-2009411563300581.6498904.sumo.cfg
 Emission file: None
 Step length: 0.2
Opening the I/O thread to SUMO
Loading configuration... done.
Error processing event.
Traceback (most recent call last):
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/tune/trial_runner.py", line 261, in _process_events
    result = self.trial_executor.fetch_result(trial)
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/tune/ray_trial_executor.py", line 211, in fetch_result
    result = ray.get(trial_future[0])
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/worker.py", line 2386, in get
    raise value
ray.worker.RayTaskError: ray_PPOAgent:train() (pid=23680, host=fortiss-n-065)
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/rllib/agents/agent.py", line 279, in train
    result = Trainable.train(self)
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/tune/trainable.py", line 146, in train
    result = self._train()
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/rllib/agents/ppo/ppo.py", line 101, in _train
    fetches = self.optimizer.step()
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/rllib/optimizers/multi_gpu_optimizer.py", line 125, in step
    self.num_envs_per_worker, self.train_batch_size)
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/rllib/optimizers/rollout.py", line 28, in collect_samples
    next_sample = ray.get(fut_sample)
ray.worker.RayTaskError: ray_PolicyEvaluator:sample() (pid=23691, host=fortiss-n-065)
  File "/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/ray/memory_monitor.py", line 78, in raise_if_low_memory
    self.error_threshold))
ray.memory_monitor.RayOutOfMemoryError: More than 95% of the memory on node fortiss-n-065 is used (15.85 / 16.69 GB). The top 5 memory consumers are:

PID MEM  COMMAND
23680 4.57GB
ray_PPOAgent:train()
23691 4.41GB
ray_PolicyEvaluator:sample()
3016 2.04GB
/home/buechel/pycharm-community-2019.1.3/jre64/bin/java -classpath /home/buechel/pycharm-community-2
2910 1.43GB
/usr/lib/x86_64-linux-gnu/zeitgeist-fts
10938 1.14GB
/home/buechel/.conda/envs/flow/bin/python /home/buechel/.conda/envs/flow/bin/tensorboard --logdir=~/

In addition, ~4.75 GB of shared memory is currently being used by the Ray object store. You can set the object store size with the `object_store_memory` parameter when starting Ray, and the max Redis size with `redis_max_memory`.


Terminating environment
Attempting to recover trial state from last checkpoint.
I0717 22:10:45.926017 23674 store.cc:599] Disconnecting client on fd 11
2019-07-17 22:10:46,316 WARNING ppo.py:137 -- By default, observations will be normalized with MeanStdFilter
Loading configuration... done.
Success.
 Starting SUMO on port 47596
 Cfg file: /home/buechel/gitfolders/flowfolder/flow/core/kernel/scenario/debug/cfg/stabilizing_open_network_merges_20190717-2210461563394246.3175511.sumo.cfg
 Emission file: None
 Step length: 0.2
Opening the I/O thread to SUMO
Loading configuration... done.
Connecting traci on port 47596
2019-07-17 22:10:47,401 INFO policy_evaluator.py:262 -- Creating policy evaluation worker 0 on CPU (please ignore any CUDA init errors)
2019-07-17 22:10:47.413078: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
/home/buechel/.conda/envs/flow/lib/python3.5/site-packages/tensorflow/python/ops/gradients_impl.py:100: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
2019-07-17 22:10:48,445 INFO multi_gpu_optimizer.py:74 -- LocalMultiGPUOptimizer devices ['/cpu:0']
== Status ==
Using FIFO scheduling algorithm.
Resources requested: 2/2 CPUs, 0/1 GPUs
Memory usage on this node: 15.6/16.7 GB: ***LOW MEMORY*** less than 10% of the memory on this node is available for use. This can cause unexpected crashes. Consider reducing the memory used by your application or reducing the Ray object store size by setting `object_store_memory` when calling `ray.init`.
Result logdir: /home/buechel/ray_results/stabilizing_open_network_merges
RUNNING trials:
 - PPO_WaveAttenuationMergePOEnv-v0_0:  RUNNING, 1 failures: /home/buechel/ray_results/stabilizing_open_network_merges/PPO_WaveAttenuationMergePOEnv-v0_0_2019-07-16_20-09-29cmzjqzle/error_2019-07-17_22-10-45.txt [pid=23680], 83367 s, 120 iter, 4512000 ts, 346 rew
```

## Versions

- OS: ubuntu 16.04 LTS 64 Bit
- Flow version: 0.3.1
- Flow commit number: c8ea2b0cd35ecc811dcdd07b51d34a0766a0ed86

**Comment by buechelmartin** [comment]
Additional info: 
Installed flow following the installation guidelines using anaconda.

Ray Version: 0.6.1

**Comment by cathywu** [comment]
Is it possible that your machine is indeed out of memory, as the error suggests? Can you try to see if you still get the same error with something like 10 rollouts instead of 50 or using an instance with more memory? For reference, we typically use instances with 30 GB memory.

**Comment by buechelmartin** [comment]
My machine has 16GB RAM, so yes that's the limit.
The question is why it is eating up all that memory. I would expect the memory usage being constant over the iterations? So if it can do the first 120 iterations, it should be able to do more as well? 

At the time of writing this, the very same experiment keeps running now with N_CPU = 4 instead of 1 as when the error above happened. It just reached checkpoint 260. This is the very first time I see it running for such a long time, most of the time all the processes just go into sleep and the experiment doesn't continue without any error message - trying to nail this one down too, and happy to open another issue to prevent others loosing their time with this.

Anyway, if you could confirm that 
1) you agree the memory usage of all the processes running from the flow experiment should not be dramatically growing over the number of iterations and 
2) the memory usage should not be less when running more workers in parallel
I will keep searching why some other processes suddenly start peaking in their memory need over night and close this issue.
That is, after my recent observations, the most likely reason why the process stops.
Sorry for bothering you then and thanks for the support!

P.S.: if you have seen the behaviour of all the processes falling asleep as well, this is happening very often and hindering me from getting any valuable training results. Hints very much appreciated!

**Comment by cathywu** [comment]
Interesting. Yes, thanks for clarifying. The memory usage should not be dramatically increasing during the run of a flow experiment. If you have identified a memory leak in flow, we'd be very interested. Is it possible that old copies of the environment are sticking around in memory between iterations?

I personally have not had issues with processes falling asleep. We are aware of some recent ray autoscaler issues and have run into process issues when using the autoscaler.

**Comment by eugenevinitsky** [comment]
Hi @buechelmartin,

Do you see the processes fall asleep right before Ray prints out it's status report on the average mean, reward, variance etc.? It's possible that one of the rollouts is taking much longer than the other so most of the rollouts complete, go to sleep. The pattern you'd see if this is the case is only one process running, then all of the processes resuming all at once.

**Comment by eugenevinitsky** [comment]
Also, can you try decreasing the total number of rollouts? This would help diagnose if it's an error due to the total batches from the rollouts eating up all of your memory i.e. keeping all of the information from the rollouts in memory smoehow winds up taking up too much memory.

**Comment by buechelmartin** [comment]
@cathywu 
I was fearing that this is a memory leak when opening this issue, since I sometimes got this error message too on another workstation with 64GB RAM. But on the occurence from the log above, I believe today this was a combination of a) flow already eating up most of the memory b) this weird zeitgeist-fts process peaking at the same time as c) Tensorboard peaking in memory usage.
I will report if I see any other strange things hinting to a memory leak.

@eugenevinitsky  
The processes all fall asleep and never recover. I believe this is something different from the MemoryError and will quickly open another issue, so it will be easier for others to track.

I will close this for the moment and only reopen if things start pointing to some issue within flow/rrlib/ray.

**Comment by GoingMyWay** [comment]
@buechelmartin Hi, is this issue fixed?

**Comment by buechelmartin** [comment]
Unfortunately I had to give up my work with flow, so no update possible. Issue was coming back every now and then until I stopped trying. But not quite reproducible.

**Comment by GoingMyWay** [comment]
> Unfortunately I had to give up my work with flow, so no update possible. Issue was coming back every now and then until I stopped trying. But not quite reproducible.

This issue can be reproducible in my project, but I cannot find a reasonable method to fix it.
